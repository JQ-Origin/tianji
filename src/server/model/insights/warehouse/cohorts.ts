import { prisma } from '../../_client.js';

export function getAllCohorts(workspaceId: string) {
  return prisma.warehouseCohorts.findMany({
    where: {
      workspaceId,
    },
  });
}
