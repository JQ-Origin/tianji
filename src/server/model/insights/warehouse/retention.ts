import { Prisma } from '@prisma/client';
import {
  findWarehouseApplication,
  WarehouseWideTableInsightsApplication,
} from './utils.js';
import dayjs from 'dayjs';
import { env } from '../../../utils/env.js';

export async function buildRetentionQuery(
  workspaceId: string,
  warehouseWideApplicationId: string,
  startAt: number,
  endAt: number
): Promise<Prisma.Sql> {
  const { sql, raw } = Prisma;
  const userApplicationId = env.insights.warehouse.retention.userApplicationId;

  // Only support wideTable application for retention
  const app = (await findWarehouseApplication(
    workspaceId,
    warehouseWideApplicationId
  )) as WarehouseWideTableInsightsApplication | undefined;

  if (!app || app.type !== 'wideTable') {
    throw new Error(
      `Retention only supports warehouse wideTable application, got: ${app?.type ?? 'unknown'}`
    );
  }

  const eventTableName = app.tableName;
  const eventUserIdField = app.distinctField;

  // Resolve user application (from env) for cohort definition
  if (!userApplicationId) {
    throw new Error(
      'env.insights.warehouse.retention.userApplicationId is not set'
    );
  }
  const userApp = (await findWarehouseApplication(
    workspaceId,
    userApplicationId
  )) as WarehouseWideTableInsightsApplication | undefined;
  if (!userApp || userApp.type !== 'wideTable') {
    throw new Error(
      `User application for retention must be a wideTable, got: ${userApp?.type ?? 'unknown'}`
    );
  }

  // Helpers to build date expr and range for a specific application/alias
  const buildActivityDateExprForApp = (
    application: WarehouseWideTableInsightsApplication,
    alias: string
  ): Prisma.Sql => {
    if (application.dateBasedCreatedAtField) {
      return sql`"${raw(alias)}"."${raw(application.dateBasedCreatedAtField)}"`;
    }

    const field = `"${alias}"."${application.createdAtField}"`;
    const type = application.createdAtFieldType ?? 'timestampMs';

    if (type === 'timestamp') {
      return sql`DATE(FROM_UNIXTIME(${raw(field)}))`;
    }
    if (type === 'timestampMs') {
      return sql`DATE(FROM_UNIXTIME(${raw(field)} / 1000))`;
    }
    if (type === 'date') {
      return sql`DATE(${raw(field)})`;
    }
    if (type === 'datetime') {
      return sql`DATE(${raw(field)})`;
    }
    return sql`DATE(FROM_UNIXTIME(${raw(field)} / 1000))`;
  };

  const buildDateRangeConditionForApp = (
    application: WarehouseWideTableInsightsApplication,
    alias: string
  ): Prisma.Sql => {
    const hasDateBased = Boolean(application.dateBasedCreatedAtField);
    const type = application.createdAtFieldType ?? 'timestampMs';
    const field = hasDateBased
      ? sql`"${raw(alias)}"."${raw(application.dateBasedCreatedAtField!)}"`
      : sql`"${raw(alias)}"."${raw(application.createdAtField)}"`;

    if (hasDateBased) {
      const startTime = dayjs(startAt).startOf('day').format('YYYY-MM-DD');
      const endTime = dayjs(endAt).endOf('day').format('YYYY-MM-DD');
      return sql`${field} BETWEEN ${startTime} AND ${endTime}`;
    }

    if (type === 'timestamp') {
      const startTime = String(dayjs(startAt).unix());
      const endTime = String(dayjs(endAt).unix());
      return sql`${field} BETWEEN ${startTime} AND ${endTime}`;
    }
    if (type === 'timestampMs') {
      const startTime = String(dayjs(startAt).valueOf());
      const endTime = String(dayjs(endAt).valueOf());
      return sql`${field} BETWEEN ${startTime} AND ${endTime}`;
    }
    if (type === 'date') {
      const startTime = dayjs(startAt).startOf('day').format('YYYY-MM-DD');
      const endTime = dayjs(endAt).endOf('day').format('YYYY-MM-DD');
      return sql`${field} BETWEEN ${startTime} AND ${endTime}`;
    }
    if (type === 'datetime') {
      const startTime = dayjs(startAt).utc().format('YYYY-MM-DD HH:mm:ss');
      const endTime = dayjs(endAt).utc().format('YYYY-MM-DD HH:mm:ss');
      return sql`${field} BETWEEN ${startTime} AND ${endTime}`;
    }

    const startTime = String(dayjs(startAt).valueOf());
    const endTime = String(dayjs(endAt).valueOf());
    return sql`${field} BETWEEN ${startTime} AND ${endTime}`;
  };

  const query = sql`
WITH "first_seen" AS (
  SELECT
    "${raw(userApp.distinctField)}" AS "user_id",
    MIN(${buildActivityDateExprForApp(userApp, 'u')}) AS "cohort_date"
  FROM "${raw(userApp.tableName)}" AS "u"
  WHERE ${buildDateRangeConditionForApp(userApp, 'u')}
  GROUP BY "${raw(userApp.distinctField)}"
)
SELECT
  "fs"."cohort_date" AS "date",
  COUNT(DISTINCT "fs"."user_id") AS "cohort_size",
  COUNT(DISTINCT CASE WHEN DATEDIFF(${buildActivityDateExprForApp(app, 'e')}, "fs"."cohort_date") = 1 THEN "fs"."user_id" END) AS "d1",
  COUNT(DISTINCT CASE WHEN DATEDIFF(${buildActivityDateExprForApp(app, 'e')}, "fs"."cohort_date") = 3 THEN "fs"."user_id" END) AS "d3",
  COUNT(DISTINCT CASE WHEN DATEDIFF(${buildActivityDateExprForApp(app, 'e')}, "fs"."cohort_date") = 5 THEN "fs"."user_id" END) AS "d5",
  COUNT(DISTINCT CASE WHEN DATEDIFF(${buildActivityDateExprForApp(app, 'e')}, "fs"."cohort_date") = 7 THEN "fs"."user_id" END) AS "d7",
  COUNT(DISTINCT CASE WHEN DATEDIFF(${buildActivityDateExprForApp(app, 'e')}, "fs"."cohort_date") = 14 THEN "fs"."user_id" END) AS "d14"
FROM "first_seen" AS "fs"
LEFT JOIN "${raw(eventTableName)}" AS "e" ON "e"."${raw(eventUserIdField)}" = "fs"."user_id" AND ${buildDateRangeConditionForApp(app, 'e')}
GROUP BY 1
ORDER BY 1 DESC`;

  return query;
}
