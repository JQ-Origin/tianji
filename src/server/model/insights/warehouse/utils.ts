import { z } from 'zod';
import { createPool, Pool } from 'mysql2/promise';
import { env } from '../../../utils/env.js';
import { Prisma } from '@prisma/client';

// MySQL date formats for different time units
export const MYSQL_DATE_FORMATS = {
  minute: '%Y-%m-%d %H:%i:00',
  hour: '%Y-%m-%d %H:00:00',
  day: '%Y-%m-%d',
  month: '%Y-%m-01',
  year: '%Y-01-01',
} as const;

export const dateTypeSchema = z.enum([
  'timestamp', // for example: 1739203200
  'timestampMs', // for example: 1739203200000
  'date', // for example: 2025-08-01
  'datetime', // for example: 2025-08-01 00:00:00
]);

const warehouseLongTableInsightsApplicationSchema = z.object({
  databaseUrl: z.string().optional(),
  name: z.string(),
  type: z.literal('longTable'), // long table
  eventTable: z.object({
    name: z.string(),
    eventNameField: z.string(),
    createdAtField: z.string(),
    createdAtFieldType: dateTypeSchema.default('timestampMs'),
    dateBasedCreatedAtField: z.string().optional(), // for improve performance, treat as date type
  }),
  eventParametersTable: z.object({
    name: z.string(),
    eventNameField: z.string(),
    paramsNameField: z.string(),
    paramsValueField: z.string(),
    paramsValueNumberField: z.string().optional(),
    paramsValueStringField: z.string().optional(),
    paramsValueDateField: z.string().optional(),
    createdAtField: z.string(),
    createdAtFieldType: dateTypeSchema.default('timestampMs'),
    dateBasedCreatedAtField: z.string().optional(), // for improve performance, treat as date type
  }),
});

const warehouseWideTableInsightsApplicationSchema = z.object({
  databaseUrl: z.string().optional(),
  name: z.string(),
  type: z.literal('wideTable'), // wide table
  tableName: z.string(),
  fields: z.array(
    z.object({
      name: z.string(),
      type: z.string().default('string'),
    })
  ),
  distinctField: z.string(),
  createdAtField: z.string(),
  createdAtFieldType: dateTypeSchema.default('timestampMs'),
  dateBasedCreatedAtField: z.string().optional(), // for improve performance, treat as date type
});

export const warehouseInsightsApplicationSchema = z.union([
  warehouseLongTableInsightsApplicationSchema,
  warehouseWideTableInsightsApplicationSchema,
]);

export type WarehouseInsightsApplication = z.infer<
  typeof warehouseInsightsApplicationSchema
>;

export type WarehouseLongTableInsightsApplication = z.infer<
  typeof warehouseLongTableInsightsApplicationSchema
>;

export type WarehouseWideTableInsightsApplication = z.infer<
  typeof warehouseWideTableInsightsApplicationSchema
>;

let applications: WarehouseInsightsApplication[] | null = null;
export function getWarehouseApplications() {
  if (!applications) {
    applications = warehouseInsightsApplicationSchema
      .array()
      .parse(JSON.parse(env.insights.warehouse.applicationsJson || '[]'));
  }

  return applications;
}

export function findWarehouseApplication(name: string) {
  return getWarehouseApplications().find((app) => app.name === name);
}

const connections = new Map<string, Pool>();
export function getWarehouseConnection(url = env.insights.warehouse.url) {
  if (!env.insights.warehouse.enable) {
    throw new Error('Warehouse is not enabled');
  }

  if (!url) {
    throw new Error('Warehouse url is not set');
  }

  let connection = connections.get(url);
  if (!connection) {
    connection = createPool(url);
    connections.set(url, connection);
  }

  return connection;
}

/**
 * Execute a data warehouse query
 * @param applicationId Application ID
 * @param sql Query SQL
 * @returns Query result
 */
export async function executeWarehouseQuery(
  applicationId: string,
  sql: Prisma.Sql
): Promise<any[]> {
  const application = findWarehouseApplication(applicationId);
  if (!application) {
    throw new Error(`Application ${applicationId} not found`);
  }

  const connection = getWarehouseConnection(application.databaseUrl);

  const [rows] = await connection.query(
    sql.sql.replaceAll('"', '`'), // avoid mysql and pg sql syntax error about double quote
    sql.values
  );

  return rows as any[];
}
