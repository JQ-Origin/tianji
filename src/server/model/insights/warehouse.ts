/**
 * Custom Warehouse for Insights
 */

import { Prisma } from '@prisma/client';
import { InsightsSqlBuilder } from './shared.js';
import { createPool, Pool } from 'mysql2/promise';
import { insightsQuerySchema } from '../../utils/schema.js';
import { z } from 'zod';
import dayjs from 'dayjs';
import { processGroupedTimeSeriesData } from './utils.js';
import { env } from '../../utils/env.js';
import { FilterInfoType, FilterInfoValue } from '@tianji/shared';

// MySQL date formats for different time units
const MYSQL_DATE_FORMATS = {
  minute: '%Y-%m-%d %H:%i:00',
  hour: '%Y-%m-%d %H:00:00',
  day: '%Y-%m-%d',
  month: '%Y-%m-01',
  year: '%Y-01-01',
} as const;

const dateTypeSchema = z.enum([
  'timestamp', // for example: 1739203200
  'timestampMs', // for example: 1739203200000
  'date', // for example: 2025-08-01
  'datetime', // for example: 2025-08-01 00:00:00
]);

export const warehouseInsightsApplicationSchema = z.object({
  name: z.string(),
  type: z.enum(['longTable', 'wideTable']).default('longTable'), // long table or wide table, TODO: implement wide table support
  eventTable: z.object({
    name: z.string(),
    eventNameField: z.string(),
    createdAtField: z.string(),
    createdAtFieldType: dateTypeSchema.default('timestampMs'),
    dateBasedCreatedAtField: z.string().optional(), // for improve performance, treat as date type
  }),
  eventParametersTable: z.object({
    name: z.string(),
    paramsNameField: z.string(),
    paramsValueField: z.string(),
    paramsValueNumberField: z.string().optional(),
    paramsValueStringField: z.string().optional(),
    paramsValueDateField: z.string().optional(),
    createdAtField: z.string(),
    createdAtFieldType: dateTypeSchema.default('timestampMs'),
    dateBasedCreatedAtField: z.string().optional(), // for improve performance, treat as date type
  }),
});

export type WarehouseInsightsApplication = z.infer<
  typeof warehouseInsightsApplicationSchema
>;

let applications: WarehouseInsightsApplication[] | null = null;
export function getWarehouseApplications() {
  if (!applications) {
    applications = warehouseInsightsApplicationSchema
      .array()
      .parse(JSON.parse(env.insights.warehouse.applicationsJson || '[]'));
  }

  return applications;
}

export class WarehouseInsightsSqlBuilder extends InsightsSqlBuilder {
  private connection: Pool | null = null;

  private getConnection(): Pool {
    if (!env.insights.warehouse.enable || !env.insights.warehouse.url) {
      throw new Error('Warehouse is not enabled');
    }

    if (!this.connection) {
      this.connection = createPool(env.insights.warehouse.url);
    }

    return this.connection;
  }

  getApplication(): WarehouseInsightsApplication {
    const name = this.query.insightId;

    const application = getWarehouseApplications().find(
      (app) => app.name === name
    );

    if (!application) {
      throw new Error(`Application ${name} not found`);
    }

    return application;
  }

  private enableDateBasedOptimizedEventTable() {
    const eventTable = this.getEventTable();
    const startAt = this.query.time.startAt;
    const endAt = this.query.time.endAt;

    return (
      !!eventTable.dateBasedCreatedAtField &&
      dayjs(endAt).diff(startAt, 'day') >= 1
    );
  }

  private enableDateBasedOptimizedEventParametersTable() {
    const eventParametersTable = this.getEventParametersTable();
    const startAt = this.query.time.startAt;
    const endAt = this.query.time.endAt;

    return (
      !!eventParametersTable.dateBasedCreatedAtField &&
      dayjs(endAt).diff(startAt, 'day') >= 1
    );
  }

  private getEventTable() {
    return this.getApplication().eventTable;
  }

  private getEventParametersTable() {
    return this.getApplication().eventParametersTable;
  }

  getTableName() {
    return this.getEventTable().name;
  }

  private getValueField(type: FilterInfoType): Prisma.Sql {
    const eventParametersTable = this.getEventParametersTable();

    const valueField =
      type === 'number'
        ? Prisma.sql`"${Prisma.raw(eventParametersTable.name)}"."${Prisma.raw(eventParametersTable.paramsValueNumberField ?? eventParametersTable.paramsValueField)}"`
        : type === 'string'
          ? Prisma.sql`"${Prisma.raw(eventParametersTable.name)}"."${Prisma.raw(eventParametersTable.paramsValueStringField ?? eventParametersTable.paramsValueField)}"`
          : type === 'date'
            ? Prisma.sql`"${Prisma.raw(eventParametersTable.name)}"."${Prisma.raw(eventParametersTable.paramsValueDateField ?? eventParametersTable.paramsValueField)}"`
            : Prisma.sql`"${Prisma.raw(eventParametersTable.name)}"."${Prisma.raw(eventParametersTable.paramsValueField)}"`;

    return valueField;
  }

  private buildFilterQueryOperator(
    type: FilterInfoType,
    operator: string,
    value: FilterInfoValue | null
  ) {
    const valueField = this.getValueField(type);

    return this.buildCommonFilterQueryOperator(
      type,
      operator,
      value,
      valueField
    );
  }

  protected getDateQuery(
    field: string,
    unit: string,
    timezone: string,
    type: z.infer<typeof dateTypeSchema> = 'timestampMs'
  ): Prisma.Sql {
    // Check if unit exists in MYSQL_DATE_FORMATS before using it
    if (!(unit in MYSQL_DATE_FORMATS)) {
      throw new Error(`Invalid date unit: ${unit}`);
    }

    const format = MYSQL_DATE_FORMATS[unit as keyof typeof MYSQL_DATE_FORMATS];
    let timeExpression: Prisma.Sql;

    switch (type) {
      case 'timestamp':
        // Unix timestamp in seconds - convert to datetime first
        timeExpression = Prisma.sql`FROM_UNIXTIME(${Prisma.raw(field)})`;
        break;
      case 'timestampMs':
        // Unix timestamp in milliseconds - convert to seconds first, then to datetime
        timeExpression = Prisma.sql`FROM_UNIXTIME(${Prisma.raw(field)} / 1000)`;
        break;
      case 'date':
        // Date string format (YYYY-MM-DD) - convert to datetime
        timeExpression = Prisma.sql`STR_TO_DATE(${Prisma.raw(field)}, '%Y-%m-%d')`;
        break;
      case 'datetime':
        // Datetime string format (YYYY-MM-DD HH:MM:SS) - use directly
        timeExpression = Prisma.sql`STR_TO_DATE(${Prisma.raw(field)}, '%Y-%m-%d %H:%i:%s')`;
        break;
      default:
        // Default to timestampMs for backwards compatibility
        timeExpression = Prisma.sql`FROM_UNIXTIME(${Prisma.raw(field)} / 1000)`;
        break;
    }

    return Prisma.sql`DATE_FORMAT(CONVERT_TZ(${timeExpression}, '+00:00', ${timezone}), '${Prisma.raw(format)}')`;
  }

  protected buildOptimizedDateRangeQuery(): Prisma.Sql {
    const { time } = this.query;
    const { startAt, endAt } = time;

    const eventTable = this.getEventTable();
    const enableDateBasedOptimizedEventTable =
      this.enableDateBasedOptimizedEventTable();
    const type = eventTable.createdAtFieldType;

    let startTime = String(dayjs(startAt).valueOf());
    let endTime = String(dayjs(endAt).valueOf());

    let field = enableDateBasedOptimizedEventTable
      ? `"${eventTable.name}"."${eventTable.dateBasedCreatedAtField ?? eventTable.createdAtField}"`
      : `"${eventTable.name}"."${eventTable.createdAtField}"`;

    if (this.enableDateBasedOptimizedEventTable()) {
      // if enable date based optimized event table,
      // and the date range is more than 1 day,
      // we need to use the date based created at field
      startTime = String(dayjs(startAt).startOf('day').format('YYYY-MM-DD'));
      endTime = String(dayjs(endAt).endOf('day').format('YYYY-MM-DD'));
    } else {
      if (type === 'timestamp') {
        // Unix timestamp in seconds
        startTime = String(dayjs(startAt).unix());
        endTime = String(dayjs(endAt).unix());
      }
      if (type === 'date') {
        // Date string format (YYYY-MM-DD)
        startTime = dayjs(startAt).startOf('day').format('YYYY-MM-DD');
        endTime = dayjs(endAt).endOf('day').format('YYYY-MM-DD');
      }
      if (type === 'datetime') {
        // Datetime string format (YYYY-MM-DD HH:mm:ss)
        startTime = dayjs(startAt).utc().format('YYYY-MM-DD HH:mm:ss');
        endTime = dayjs(endAt).utc().format('YYYY-MM-DD HH:mm:ss');
      }
    }

    return Prisma.sql`${Prisma.raw(field)} BETWEEN ${startTime} AND ${endTime}`;
  }

  protected getOptimizedDateQuery(): Prisma.Sql {
    const { time } = this.query;
    const { unit, timezone = 'UTC' } = time;
    const eventTable = this.getEventTable();
    const enableDateBasedOptimizedEventTable =
      this.enableDateBasedOptimizedEventTable();
    const type = enableDateBasedOptimizedEventTable
      ? eventTable.dateBasedCreatedAtField
        ? 'date'
        : eventTable.createdAtFieldType
      : eventTable.createdAtFieldType;
    const field = enableDateBasedOptimizedEventTable
      ? `"${eventTable.name}"."${eventTable.dateBasedCreatedAtField ?? eventTable.createdAtField}"`
      : `"${eventTable.name}"."${eventTable.createdAtField}"`;

    // Check if unit exists in MYSQL_DATE_FORMATS before using it
    if (!(unit in MYSQL_DATE_FORMATS)) {
      throw new Error(`Invalid date unit: ${unit}`);
    }

    const format = MYSQL_DATE_FORMATS[unit as keyof typeof MYSQL_DATE_FORMATS];
    let timeExpression: Prisma.Sql;

    switch (type) {
      case 'timestamp':
        // Unix timestamp in seconds - convert to datetime first
        timeExpression = Prisma.sql`FROM_UNIXTIME(${Prisma.raw(field)})`;
        break;
      case 'timestampMs':
        // Unix timestamp in milliseconds - convert to seconds first, then to datetime
        timeExpression = Prisma.sql`FROM_UNIXTIME(${Prisma.raw(field)} / 1000)`;
        break;
      case 'date':
        // Date string format (YYYY-MM-DD) - convert to datetime
        timeExpression = Prisma.sql`STR_TO_DATE(${Prisma.raw(field)}, '%Y-%m-%d')`;
        break;
      case 'datetime':
        // Datetime string format (YYYY-MM-DD HH:MM:SS) - use directly
        timeExpression = Prisma.sql`STR_TO_DATE(${Prisma.raw(field)}, '%Y-%m-%d %H:%i:%s')`;
        break;
      default:
        // Default to timestampMs for backwards compatibility
        timeExpression = Prisma.sql`FROM_UNIXTIME(${Prisma.raw(field)} / 1000)`;
        break;
    }

    return Prisma.sql`DATE_FORMAT(CONVERT_TZ(${timeExpression}, '+00:00', ${timezone}), '${Prisma.raw(format)}')`;
  }

  protected buildOptimizedDateQuerySql(): Prisma.Sql {
    const { time } = this.query;
    const { unit, timezone = 'UTC' } = time;
    const eventTable = this.getEventTable();

    if (
      this.enableDateBasedOptimizedEventTable() &&
      eventTable.dateBasedCreatedAtField
    ) {
      const field = `"${eventTable.name}"."${eventTable.dateBasedCreatedAtField}"`;
      const format =
        MYSQL_DATE_FORMATS[unit as keyof typeof MYSQL_DATE_FORMATS];

      return Prisma.sql`DATE_FORMAT(${Prisma.raw(field)}, '${Prisma.raw(format)}') date`;
    } else {
      return Prisma.sql`${this.getDateQuery(
        `"${eventTable.name}"."${eventTable.createdAtField}"`,
        unit,
        timezone,
        eventTable.createdAtFieldType
      )} date`;
    }
  }

  protected buildDateQuerySql(): Prisma.Sql {
    return this.buildOptimizedDateQuerySql();
  }

  buildSelectQueryArr() {
    const { metrics } = this.query;
    const eventTable = this.getEventTable();

    return metrics.map((item) => {
      if (item.math === 'events') {
        if (item.name === '$all_event') {
          return Prisma.sql`count(1) as "$all_event"`;
        }

        return Prisma.sql`sum(case WHEN "${Prisma.raw(eventTable.name)}"."${Prisma.raw(eventTable.eventNameField)}" = ${item.name} THEN 1 ELSE 0 END) as ${Prisma.raw(`"${item.name}"`)}`;
      }

      return null;
    });
  }

  buildGroupSelectQueryArr() {
    const { groups } = this.query;
    let groupSelectQueryArr: Prisma.Sql[] = [];
    if (groups.length > 0) {
      for (const g of groups) {
        if (!g.customGroups) {
          groupSelectQueryArr.push(
            Prisma.sql`${this.getValueField(g.type)} as "%${Prisma.raw(g.value)}"`
          );
        } else if (g.customGroups && g.customGroups.length > 0) {
          for (const cg of g.customGroups) {
            groupSelectQueryArr.push(
              Prisma.sql`${this.buildFilterQueryOperator(
                g.type,
                cg.filterOperator,
                cg.filterValue
              )} as "%${Prisma.raw(`${g.value}|${cg.filterOperator}|${cg.filterValue}`)}"`
            );
          }
        }
      }
    }
    return groupSelectQueryArr;
  }

  buildWhereQueryArr() {
    const { metrics } = this.query;
    const eventTable = this.getEventTable();

    const whereConditions = [
      // date
      this.buildOptimizedDateRangeQuery(),

      // event name
      Prisma.join(
        metrics.map((item) => {
          if (item.name === '$all_event') {
            return Prisma.sql`1 = 1`;
          }

          return Prisma.sql`"${Prisma.raw(eventTable.name)}"."${Prisma.raw(eventTable.eventNameField)}" = ${item.name}`;
        }),
        ' OR ',
        '(',
        ')'
      ),
    ];

    return whereConditions;
  }

  async executeQuery(sql: Prisma.Sql): Promise<any[]> {
    const connection = this.getConnection();

    const [rows] = await connection.query(
      sql.sql.replaceAll('"', '`'), // avoid mysql and pg sql syntax error about double quote
      sql.values
    );

    return rows as any[];
  }
}

export async function insightsWarehouse(
  query: z.infer<typeof insightsQuerySchema>,
  context: { timezone: string }
) {
  const builder = new WarehouseInsightsSqlBuilder(query, context);
  const sql = builder.build();

  const data = await builder.executeQuery(sql);

  const result = processGroupedTimeSeriesData(query, context, data);

  return result;
}
