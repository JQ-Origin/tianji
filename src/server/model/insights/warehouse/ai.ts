import { tool, ToolSet } from 'ai';
import { env } from '../../../utils/env.js';
import { getWarehouseConnection } from './utils.js';
import z from 'zod';
import { logger } from '../../../utils/logger.js';

export interface WarehouseTableColumnMeta {
  field: string;
  type: string;
  allowNull: boolean;
  key: string | null;
  default: string | number | null;
  extra: string | null;
}

export interface WarehouseTableMeta {
  table: string;
  columns: WarehouseTableColumnMeta[];
}

const cachedWarehouseTables = new Map<string, WarehouseTableMeta[]>();
export async function getWarehouseTables(
  url = env.insights.warehouse.url
): Promise<WarehouseTableMeta[]> {
  if (!url) {
    throw new Error('Warehouse url is not set');
  }

  if (cachedWarehouseTables.has(url)) {
    return cachedWarehouseTables.get(url)!;
  }

  const connection = getWarehouseConnection(url);
  const [res, fields] = await connection.query('SHOW TABLES');

  if (fields.length > 1) {
    throw new Error('Multiple tables found, not supported yet.');
  }

  const tableNames = Array.from<Record<string, string>>(res as any).map(
    (row) => row[fields[0].name]
  );

  const tables = await Promise.all<WarehouseTableMeta>(
    tableNames.map(async (table) => {
      const [res] = await connection.query(`SHOW COLUMNS FROM ${table}`);

      return {
        table,
        columns: Array.from<Record<string, string>>(res as any).map(
          (row): WarehouseTableColumnMeta => {
            return {
              field: row['Field'],
              type: row['Type'],
              allowNull: row['Null'] === 'YES' ? true : false,
              key: (row['Key'] ?? null) as string | null,
              default: (row['Default'] ?? null) as string | number | null,
              extra: (row['Extra'] ?? null) as string | null,
            };
          }
        ),
      };
    })
  );

  cachedWarehouseTables.set(url, tables);

  return tables;
}
export async function getWarehouseTable(
  tableName: string
): Promise<WarehouseTableMeta | undefined> {
  const tables = await getWarehouseTables();
  return tables.find((table) => table.table === tableName);
}

export const warehouseAITools: ToolSet = {
  getWarehouseTables: tool({
    description: 'Get the warehouse tables',
    inputSchema: z.object({}),
    execute: async () => {
      return await getWarehouseTables();
    },
  }),
  getWarehouseTable: tool({
    description: 'Get the warehouse table',
    inputSchema: z.object({
      tableName: z
        .string()
        .describe('The name of the table to get the description of'),
    }),
    execute: async ({ tableName }) => {
      return await getWarehouseTable(tableName);
    },
  }),
  queryWarehouse: tool({
    description: 'Execute a warehouse query with sql',
    inputSchema: z.object({
      sql: z.string().describe('The SQL query to execute'),
    }),
    execute: async ({ sql }) => {
      const connection = getWarehouseConnection(env.insights.warehouse.url);

      logger.log('[queryWarehouse]', sql);

      const [res] = await connection.query(sql);

      logger.log('[queryWarehouse] result', res);

      return res;
    },
  }),
  createCharts: tool({
    description: 'Create charts from the warehouse data',
    inputSchema: z.object({
      data: z
        .object({})
        .passthrough()
        .array()
        .describe(
          'The data to create charts from, its should be a object array, which must be have a field named date and other field show other info count'
        ),
      type: z
        .enum(['line', 'bar', 'pie', 'area', 'scatter', 'donut'])
        .describe('The type of chart to create'),
      title: z.string().describe('The title of the chart'),
    }),
  }),
};

export const warehouseAISystemPrompt = `
You are a helpful assistant that can help with tasks related to the warehouse.
You can use the following tools to help the user:
${JSON.stringify(Object.keys(warehouseAITools))}

The user is a business owner who is interested in analyzing their warehouse data.

## Rules

- Before generating an SQL query, make sure the context includes the structure of the database tables. If it does not, please use the getWarehouseTables or getWarehouseTable tool to obtain the table structure.
- After you finish creating the SQL, please try to call the queryWarehouse tool to execute the SQL query.
- After you use the queryWarehouse tool to complete the query, please try to call the createCharts tool to create charts.
`;
