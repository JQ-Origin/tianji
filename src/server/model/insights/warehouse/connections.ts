import { prisma } from '../../_client.js';
import { getWarehouseTables } from './utils.js';

/**
 * Upsert warehouse table records by database id and connection uri.
 */
export async function upsertWarehouseTable(
  databaseId: string,
  connectionUri: string
): Promise<{ created: number; updated: number; deleted: number }> {
  return await prisma.$transaction(async (tx) => {
    const database = await tx.warehouseDatabase.findUnique({
      where: { id: databaseId },
      select: { id: true, workspaceId: true },
    });
    if (!database) {
      throw new Error('Warehouse database not found');
    }

    const [liveTables, existing] = await Promise.all([
      getWarehouseTables(connectionUri),
      tx.warehouseDatabaseTable.findMany({
        where: { workspaceId: database.workspaceId, databaseId: database.id },
        select: { id: true, name: true },
      }),
    ]);
    const nameToRecord = new Map(existing.map((r) => [r.name, r]));

    let created = 0;
    let updated = 0;
    let deleted = 0;

    for (const table of liveTables) {
      const record = nameToRecord.get(table.tableName);
      if (record) {
        await tx.warehouseDatabaseTable.update({
          where: { id: record.id },
          data: { ddl: table.ddl },
        });
        updated += 1;
      } else {
        await tx.warehouseDatabaseTable.create({
          data: {
            workspaceId: database.workspaceId,
            databaseId: database.id,
            name: table.tableName,
            ddl: table.ddl,
          },
        });
        created += 1;
      }
    }

    // Delete records that no longer exist in liveTables for this database
    const liveNames = new Set(liveTables.map((t) => t.tableName));
    const toDeleteIds = existing
      .filter((r) => !liveNames.has(r.name))
      .map((r) => r.id);

    if (toDeleteIds.length > 0) {
      const { count } = await tx.warehouseDatabaseTable.deleteMany({
        where: { id: { in: toDeleteIds } },
      });
      deleted = count;
    }

    return { created, updated, deleted };
  });
}
