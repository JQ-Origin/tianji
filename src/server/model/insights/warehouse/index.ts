import { z } from 'zod';
import { insightsQueryEventsSchema } from '../../../utils/schema.js';
import { WarehouseLongTableInsightsSqlBuilder } from './longTable.js';
import { findWarehouseApplication } from './utils.js';
import { WarehouseWideTableInsightsSqlBuilder } from './wideTable.js';

export async function queryWarehouseEvents(
  query: z.infer<typeof insightsQueryEventsSchema>,
  context: { timezone: string }
) {
  const application = await findWarehouseApplication(
    query.workspaceId,
    query.insightId
  );

  if (application?.type === 'wideTable') {
    const builder = new WarehouseWideTableInsightsSqlBuilder(query, context);
    return builder.queryEvents(query.cursor);
  }

  const builder = new WarehouseLongTableInsightsSqlBuilder(query, context);
  return builder.queryEvents(query.cursor);
}
