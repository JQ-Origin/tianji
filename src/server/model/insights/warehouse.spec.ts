import { beforeAll, describe, expect, test } from 'vitest';
import { WarehouseInsightsSqlBuilder } from './warehouse.js';
import { unwrapSQL } from '../../utils/prisma.js';
import dayjs from 'dayjs';
import { env } from '../../utils/env.js';

describe('WarehouseInsightsSqlBuilder', () => {
  const insightId = 'test'; // application name
  const insightType = 'warehouse';

  beforeAll(() => {
    env.insights.warehouse.applicationsJson = JSON.stringify([
      {
        name: 'test',
        eventTable: {
          name: 'events',
          eventNameField: 'event_name',
          createdAtField: 'event_timestamp',
        },
        eventParametersTable: {
          name: 'event_parameters',
          paramsNameField: 'event_param_key',
          paramsValueField: 'event_param_value',
          createdAtField: 'event_timestamp',
        },
      },
    ]);
  });

  test('default', () => {
    const builder = new WarehouseInsightsSqlBuilder(
      {
        insightId,
        insightType,
        workspaceId: '',
        metrics: [
          {
            name: '$all_event',
            math: 'events',
          },
        ],
        filters: [],
        time: {
          startAt: dayjs('2025-08-01').valueOf(),
          endAt: dayjs('2025-08-02').valueOf(),
          unit: 'day',
        },
        groups: [],
      },
      {
        timezone: 'UTC',
      }
    );

    const sql = builder.build();
    expect(unwrapSQL(sql)).toMatchSnapshot('sql');
  });
});
