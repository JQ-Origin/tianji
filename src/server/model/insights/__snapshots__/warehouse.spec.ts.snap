// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WarehouseInsightsSqlBuilder > default > sql 1`] = `
"select
        DATE_FORMAT(CONVERT_TZ(FROM_UNIXTIME("events"."event_timestamp" / 1000), '+00:00', 'UTC'), '%Y-%m-%d') date , count(1) as "$all_event"
      from "events" 
      where "events"."event_timestamp" BETWEEN '1753977600000' AND '1754064000000' AND (1 = 1)
      group by 1
      order by 1 desc
      limit 1000"
`;
